File: 백준/Diamond/16214. N과 M/N과 M.cc
#include<stdio.h>
#define ll long long
ll pow(int x, int y, int p){
	x%=p;
	ll t;
	if (y==0){
		return 1;
	}
	if (y==1){
		return x%p;
	}
	t=pow(x,y/2,p);
	if (y%2){
		return (t*t)%p*x%p;
	}
	else{
		return (t*t)%p;
	}
}

ll phi(int n){
	ll res=n;
	for (int i=2;i*i<=n;i++){
		if (n%i==0){
			res-=(res/i);
			while (n%i==0){
				(n/=i);
			}
		}
	}
	if (n>1){
		res-=res/n;
	}
	return res;
}

int sol(int n, int m){
	int s=1;
	ll p[100]={m};
	for (int i=1;i<100;i++){
		p[i]=phi(p[i-1]);
		if (p[i]==1){
			s=i;
			break;
		}
	}
	int arr[100]={0};
	arr[s]=1;
	for (int i=s-1;i>0;i--){
		arr[i]=pow(n,arr[i+1],p[i])+p[i];
	}
	arr[0]=pow(n,arr[1],m);
	return arr[0];
}

int main(){
	int n,m,a;
	scanf("%d",&a);
	for (int i=0;i<a;i++){
		scanf("%d %d",&n,&m);
		if (m==1){
			printf("0\n");
		}
		else if (m==2){
			printf("%d\n",n%2);
		}
		else if (n==1){
			printf("1\n");
		}
		else{
			printf("%d\n",sol(n,m));
		}
	}
}

File: 백준/Diamond/16214. N과 M/N과 M.py
import sys
input=sys.stdin.readline
import math

def pow(x,y,p):
    res=1
    while y>0:
        if y%2:
            res*=x
            res%=p
        x*=x
        x%=p
        y//=2
    return res
    
def phi(n):
    res=n
    for i in range(2,math.isqrt(n)+2):
        if n%i==0:
            res-=(res//i)
            while (n%i==0):
                n//=i
    if n>1:
        res-=(res//n)
    return res

a=int(input())
def sol(n,m):
    s=1
    p=[m]*50
    for i in range(1,50):
        p[i]=phi(p[i-1])
        if p[i]==1:
            s=i
            break
    arr=[0]*32
    arr[s]=1
    for i in range(s-1,0,-1):
        arr[i]=pow(n,arr[i+1],p[i])+p[i]
    arr[0]=pow(n,arr[1],m)
    return arr[0]
    
for _ in range(a):
    n,m=map(int,input().split())
    if m==1:
        print(0)
    elif n==1:
        print(1)
    elif m==2:
        print(n%2)
    else:
        print(sol(n,m))